Relay Test App Example C++

francobithead opened the pull request.

francobithead committed:
Project Set Up

francobithead committed:
Adjust settings for plugins + A few classes to get started in

francobithead committed:
Resolved widget errors & added arrow assets and other assets for reference

francobithead committed:
Callback class set up and able to Authenticate + Update Name(Output console only)

francobithead committed:
Sign In transition to Main Menu Screen + Set up widget classes for different screens

- Transitions from Sign in -> Loading Screen to wait for response from brainCloud -> Main menu which shows the users name.

francobithead committed:
RTT responses set up, still working on functions to reflect responses to UE4

francobithead committed:
Moved RelayUserData from struct to object + adding logic to lobby widget + more

- Made a OtherMatchUserWidget class
- Adjusted functions in RelayNetworkInterface to not be implemented, rather just a normal function being called from widget classes.
- Color buttons set up
- Joined and Update members responses set up to reflect in UE4

francobithead committed:
Lobby almost done but still WIP

-> Able to add local and other members to lobby list
-> Local user is able to change color and send to everyone else in lobby
-> Local user can receive updates from other members when color is changed
WIP:
-> Local user leaving currently crashes engine(yay)
-> Start button isn't being consistent, will need to investigate.

francobithead committed:
Start and Leave buttons are working now. Next is Match screen

francobithead committed:
Match Screen Iteration: Cursors are set up

Cursors set up:
- Game can send mouse events when hovering over the game area
- Game can receive mouse events and reflect it on the game area
- Set up interface calls for Relay, System Relay, and Connect Relay to handle callbacks for Relay events.
- Made a Player Controller class to change mouse cursor
- Refactored abit for functions to be renamed more clearly of what they're doing (CreateUser instead of CreateUserAndAddToList and FindLobby instead of EnableRTTCallback)

francobithead committed:
Match Iteration Continued; Shockwaves are set up !

- Game can send Shockwaves to other members
- Game can receive Shockwaves from other members
- Colors are applied to shockwave for both local and other members in match.
- Some refactoring for input events being sent to BC and logic to move cursors from responses got moved to MatchWidget class (Function name: MoveOtherUserCursor).
- ShockwaveWidget class created for adjusting color when created.

francobithead committed:
Cancel Join Lobby button set up + Setting version number to text on all screens

francobithead committed:
Refactored SignIn widget from blueprint & Adjust shockwave visibility function completed

francobithead committed:
Dropdown menu for relay connection type set up !

francobithead committed:
Making Convert Utilities class accessible in plugin & clean up + documenting

francobithead committed:
Renaming functions + adjusting comments

davidstl commented on RelayTestAppCPP/Source/RelayTestAppCPP/RelayGameData/RelayGameInstance.cpp:
```diff
@@ -0,0 +1,122 @@
+
+void URelayGameInstance::RemoveUserFromList(FString in_profileId)
+{
+	ARelayUserData* userToRemove;
```
Initialize to null, very dangerous if profileId is not found.
```
ARelayUserData* userToRemove = nullptr;
```

davidstl commented on RelayTestAppCPP/Source/RelayTestAppCPP/RelayGameData/RelayGameInstance.cpp:
```diff
@@ -0,0 +1,122 @@
+			userToRemove = user;
+		}
+	}
+	ListOfUserObjects.Remove(userToRemove);
```
I would null check here before calling Remove, but I presume Unreal won't crash for that.

davidstl commented on RelayTestAppCPP/Source/RelayTestAppCPP/RelayGameData/RelayGameInstance.cpp:
```diff
@@ -0,0 +1,122 @@
+	{
+		if(user->ProfileID.Equals(in_ProfileID))
+		{
+			user->bShowShockwave = in_IsVisible;
```
Add `break;` after this line. You found it, pointless to continue looping

davidstl commented on RelayTestAppCPP/Source/RelayTestAppCPP/GameRelayCallback.cpp:
```diff
@@ -0,0 +1,77 @@
+	}
+	else if(serviceName == ServiceName::RTTRegistration)
+	{
+		Interface->FindLobby();
```
no `delete this` after this line?

francobithead replied:
I couldn't figure out how I was suppose to handle this other than deleting it within the serverError() function. The reason why I don't delete here is because when the user disconnects then I would get an error `EXCEPTION_ACCESS_VIOLATION` for BrainCloudRTTComms.cpp -> 450. From what I understood I need the lobby callback to be alive until the user disconnects and deregisters RTT lobby callback otherwise the whole engine crashes with that error code.

davidstl commented on RelayTestAppCPP/Source/RelayTestAppCPP/RelayNetworkInterface.cpp:
```diff
@@ -0,0 +1,442 @@
+	{
+		userCursor->RemoveFromParent();
+	}
+	GameInstance->GameWidget->MatchWidget->UserCursors.Empty();
```
My first instinct is that those widgets are all leaking now. Not destroyed, deleted, etc. Just raw pointers.
But after reading some posts and asking a friend (unreal dev for many years), apparently, this is fine here..
Unreal makes memory management very unclear.

davidstl commented on RelayTestAppCPP/Source/RelayTestAppCPP/RelayNetworkInterface.h:
```diff
@@ -0,0 +1,146 @@
+	UPROPERTY()
+	FTimerHandle DelayTimerHandle;
+
+	class GameRelayCallback* Callback;
```
You are including the file above. You don't need to prepend with keyword `class`. No need for forward declaration.

davidstl commented on RelayTestAppCPP/Source/RelayTestAppCPP/Widgets/MatchWidget.cpp:
```diff
@@ -0,0 +1,121 @@
+		{
+			user->Arrow_Image->SetVisibility(ESlateVisibility::HitTestInvisible);
+			UCanvasPanelSlot* widgetSlot = Cast<UCanvasPanelSlot>(user->Slot);
+			widgetSlot->SetPosition(in_inputPosition);
```
again should `break;` after finding what you were looking for

davidstl commented on RelayTestAppCPP/Source/RelayTestAppCPP/Widgets/MatchWidget.cpp:
```diff
@@ -0,0 +1,121 @@
+		if(cursor->UserData->ProfileID.Equals(in_profileId))
+		{
+			cursor->RemoveFromParent();
+			cursorToRemove = cursor;
```
`break;` again

davidstl reviewed (approved):
I think it looks good. It's hard to tell. Unreal's memory management has a lot of "magic" in it. I am seeing a lot of raw pointers and potential leaks here, but unreal might manage things differently if those objects were created using NewObject, CreateWidget, etc. So it's probably fine.

francobithead committed:
Adding breaks and nullptr init for variables and adding nullptr checks.

francobithead committed:
Merge branch 'develop' into RelayTestApp-C++

francobithead committed:
Adding Braincloud image

I added back the class for GameRelayCallback due to it not being a UCLASS type of class and in order to utilize UCLASS the callback has to be an object which invokes changes to the project and I'm not sure if this method would even work effectively in the end.
